File: Laboratory_7/Laboratory_3/C++/task2/prime_algorithms.cpp
#include <iostream>
#include <random>
#include <cmath>
#include <set>

#include "utils.h"

using namespace std;
using ull = unsigned long long;

bool miller_test(ull n, int t) {
    /* Вероятностный тест Миллера */
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> dist(2, n - 1);

    if (n == 2 || n == 3 || n == 5 || n == 7) return true;
    if (n < 10) return false;

    set<ull> nums;
    while ((int)nums.size() != t) {
        nums.insert(dist(gen));
    }

    for (auto& a : nums) {
        if (modPow(a, n - 1, n) != 1) {
            return false;
        }
    }

    vector<pair<int, int>> primeFactors = factorize(n - 1);
    for (auto& factor : primeFactors) {
        int q = factor.first;
        bool onlyOne = true;
        for (auto& a : nums) {
            if (modPow(a, (n - 1) / q, n) != 1) {
                onlyOne = false;
                break;
            }
        }
        if (onlyOne) {
            return false;
        }
    }
    return true;
}

bool pocklington_test(ull n, int t) {
    /* Вероятностный тест Поклингтона */
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> distR(1, 3);
    uniform_int_distribution<> distA(2, n - 1);

    if (n == 2 || n == 3 || n == 5 || n == 7) return true;
    if (n < 10) return false;

    int r = distR(gen);
    vector<pair<int, int>> primeFactors = factorize((n - 1) / r);

    set<ull> nums;
    while ((int)nums.size() != t) {
        nums.insert(distA(gen));
    }

    for (auto& a : nums) {
        if (modPow(a, n - 1, n) != 1) {
            return false;
        }
    }

    for (auto& a : nums) {
        bool noOne = true;
        for (auto& factor : primeFactors) {
            int q = factor.first;
            if (modPow(a, (n - 1) / q, n) == 1) {
                noOne = false;
                break;
            }
        }
        if (noOne) {
            return true;
        }
    }
    return false;
}

pair<ull, int> miller_prime(int num_size, vector<int> primes) {
    /* Генерация простого числа с помощью теста Миллера */
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> distQ(0, primes.size() - 1);
    uniform_int_distribution<> distA(1, 20);
    
    ull n = 1;
    int failedMillerCount = -1;
    while (!miller_test(n, 8)) {
        int attemptCount = 0;
        ull m = 1;
        set<int> uniqQ;
        while (sizeNum(m) != num_size - 1) {
            int q = primes[distQ(gen)];
            int a = distA(gen);
            if (sizeNum(m * (int)pow(q, a)) <= num_size - 1 && uniqQ.find(q) == uniqQ.end()) {
                m *= (int)pow(q, a);
                uniqQ.insert(q);
            }
            if (++attemptCount == 100 && sizeNum(m) != num_size - 1) {
                m = 1;
                attemptCount = 0;
                uniqQ.clear();
            }
        }
        n = 2 * m + 1;
        failedMillerCount++;
    }
    return {n, failedMillerCount};
}

pair<ull, int> pocklington_prime(int num_size, vector<int> primes) {
    /* Генерация простого числа с помощью теста Поклингтона*/
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> distQ(0, primes.size() - 1);
    uniform_int_distribution<> distA(1, 20);
    
    ull n = 1;
    int failedPocklingtonCount = -1;
    while (!pocklington_test(n, 8)) {
        int attemptCount = 0;
        ull f = 1;
        set<int> uniqQ;
        while (sizeNum(f) - 1 != num_size / 2) {
            int q = primes[distQ(gen)];
            int a = distA(gen);
            if (sizeNum(f * (int)pow(q, a)) - 1 <= num_size / 2) {
                f *= (int)pow(q, a);
                uniqQ.insert(q);
            }
            if (++attemptCount == 100 && sizeNum(f) - 1 != num_size / 2) {
                f = 1;
                attemptCount = 0;
                uniqQ.clear();
            }
        }
        ull r = f >> 1;
        if (r % 2) {
            r++;
        }
        n = r * f + 1;
        failedPocklingtonCount++;
    }
    return {n, failedPocklingtonCount};
}

pair<ull, int> gost_prime(int num_size, vector<int> primes) {
    /* Генерация простого числа с помощью ГОСТ Р34.10-94*/
    random_device rd;
    mt19937 gen(rd());
    uniform_real_distribution read_dist(0.0, 1.0);

    if (num_size <= 1) return {0, 0};
    if (num_size == 2) return {3, 0};

    ull p = 0;
    ull n = 0;
    ull u = 0;
    ull q = miller_prime(ceil(num_size / 2), primes).first;
    while (true) {
        double e = read_dist(gen);
        n = ceil(pow(2, num_size - 1) / q) + ceil(pow(2, num_size - 1) * e / q);
        if (n % 2) {
            n++;
        }

        u = 0;
        p = (n + u) * q + 1;
        if (p <= (ull)pow(2, num_size)) {
            break;
        }
    }

    int failedGostCount = -1;
    while (!(modPow(2, p - 1, p) == 1 && modPow(2, n + u, p) != 1)) {
        u += 2;
        p = (n + u) * q + 1;
        failedGostCount++;
    }
    failedGostCount = failedGostCount == -1 ? 0 : failedGostCount;
    return {p, failedGostCount};
}

File: Laboratory_3/GoLang/task2/primality/prime_algorithms.go
package algorithms

import (
	"math"
	"math/rand"
	"task2/utils"
)

type Pair struct {
	First  uint64
	Second int
}

func MillerTest(n uint64, t int) bool {
	if n == 2 || n == 3 || n == 5 || n == 7 {
		return true
	}
	if n < 10 {
		return false
	}

	nums := make(map[uint64]struct{})
	for len(nums) < t {
		a := uint64(rand.Intn(int(n-3)) + 2)
		nums[a] = struct{}{}
	}

	for a := range nums {
		if utils.ModPow(a, n-1, n) != 1 {
			return false
		}
	}

	factors := utils.Factorize(n - 1)
	for _, q := range factors {
		onlyOne := true
		for a := range nums {
			if utils.ModPow(a, (n-1)/uint64(q.First), n) != 1 {
				onlyOne = false
				break
			}
		}
		if onlyOne {
			return false
		}
	}
	return true
}

func PocklingtonTest(n uint64, t int) bool {
	if n == 2 || n == 3 || n == 5 || n == 7 {
		return true
	}
	if n < 10 {
		return false
	}

	r := rand.Intn(3) + 1
	factors := utils.Factorize((n - 1) / uint64(r))

	nums := make(map[uint64]struct{})
	for len(nums) < t {
		a := uint64(rand.Intn(int(n-3)) + 2)
		nums[a] = struct{}{}
	}

	for a := range nums {
		if utils.ModPow(a, n-1, n) != 1 {
			return false
		}
	}

	for a := range nums {
		noOne := true
		for _, q := range factors {
			if utils.ModPow(a, (n-1)/uint64(q.First), n) == 1 {
				noOne = false
				break
			}
		}
		if noOne {
			return true
		}
	}
	return false
}

func MillerPrime(size int, primes []int) (uint64, int) {
	var n uint64 = 1
	failCount := -1
	for !MillerTest(n, 8) {
		attempts := 0
		m := uint64(1)
		uniq := make(map[int]bool)
		for utils.SizeNum(m) != size-1 {
			q := primes[rand.Intn(len(primes))]
			a := rand.Intn(20) + 1
			if utils.SizeNum(m*uint64(math.Pow(float64(q), float64(a)))) <= size-1 && !uniq[q] {
				m *= uint64(math.Pow(float64(q), float64(a)))
				uniq[q] = true
			}
			attempts++
			if attempts == 100 && utils.SizeNum(m) != size-1 {
				m = 1
				attempts = 0
				uniq = make(map[int]bool)
			}
		}
		n = 2*m + 1
		failCount++
	}
	return n, failCount
}

func PocklingtonPrime(size int, primes []int) (uint64, int) {
	var n uint64 = 1
	failCount := -1
	for !PocklingtonTest(n, 8) {
		attempts := 0
		f := uint64(1)
		for utils.SizeNum(f)-1 != size/2 {
			q := primes[rand.Intn(len(primes))]
			a := rand.Intn(20) + 1
			if utils.SizeNum(f*uint64(math.Pow(float64(q), float64(a))))-1 <= size/2 {
				f *= uint64(math.Pow(float64(q), float64(a)))
			}
			attempts++
			if attempts == 100 && utils.SizeNum(f)-1 != size/2 {
				f = 1
				attempts = 0
			}
		}
		r := f >> 1
		if r%2 == 1 {
			r++
		}
		n = r*f + 1
		failCount++
	}
	return n, failCount
}

func GostPrime(size int, primes []int) (uint64, int) {
	if size <= 1 {
		return 0, 0
	}
	if size == 2 {
		return 3, 0
	}

	q, _ := MillerPrime(int(math.Ceil(float64(size)/2)), primes)
	var p, n, u uint64
	e := rand.Float64()
	n = uint64(math.Ceil(math.Pow(2, float64(size-1))/float64(q))) + uint64(math.Ceil(math.Pow(2, float64(size-1))*e/float64(q)))
	if n%2 != 0 {
		n++
	}
	p = (n + u) * q + 1
	for p > uint64(math.Pow(2, float64(size))) {
		e = rand.Float64()
		n = uint64(math.Ceil(math.Pow(2, float64(size-1))/float64(q))) + uint64(math.Ceil(math.Pow(2, float64(size-1))*e/float64(q)))
		if n%2 != 0 {
			n++
		}
		p = (n + u) * q + 1
	}

	failCount := -1
	for !(utils.ModPow(2, p-1, p) == 1 && utils.ModPow(2, n+u, p) != 1) {
		u += 2
		p = (n + u) * q + 1
		failCount++
	}
	if failCount == -1 {
		failCount = 0
	}
	return p, failCount
}
